src/imars3d/backend/io/data.py

```diff
@@ -0,0 +1,431 @@
+#!/usr/bin/env python3
+"""
+Data handling for imars3d.
+"""
+import re
+import param
+import multiprocessing
+import numpy as np
+import dxchange
+import tifffile
+from functools import partial
+from pathlib import Path
+from fnmatch import fnmatchcase
+from typing import Optional, Tuple, List, Callable
+from tqdm.contrib.concurrent import process_map
+
+
+# setup module level logger
+logger = param.get_logger(__name__)
+logger.name = __name__
+METADATA_DICT = {
+    65026: "ManufacturerStr:Andor",  # [ct, ob, dc]
+    65027: "ExposureTime:70.000000",  # [ct, ob, dc]
+    65068: "MotSlitHR.RBV:10.000000",  # APERTURE_HR, [ct, ob]
+    65070: "MotSlitHL.RBV:20.000000",  # APERTURE_HL, [ct, ob]
+    65066: "MotSlitVT.RBV:10.000000",  # APERTURE_VT, [ct, ob]
+    65068: "MotSlitHR.RBV:10.000000",  # APERTURE_VB, [ct, ob]
+}
+
+
+class load_data(param.ParameterizedFunction):
+    """
+    Load data with given input
+
+    Parameters
+    ---------
+    ct_files: str
+        explicit list of radiographs
+    ob_files: str
+        explicit list of open beams
+    dc_files: Optional[str]
+        explicit list of dark current
+    ct_dir: str
+        directory contains radiographs
+    ob_dir: str
+        directory contains open beams
+    dc_dir: Optional[str]
+        directory contains dark currents
+    ct_fnmatch: Optional[str]
+        Unix shells-style wild card (``*.tiff``) for selecting radiographs
+    ob_fnmatch: Optional[str]
+        Unix shells-style wild card (``*.tiff``) for selecting open beams
+    dc_fnmatch: Optional[str]
+        Unix shells-style wild card (``*.tiff``) for selecting dark current
+    max_workers: Optional[int]
+        maximum number of processes allowed during loading, default to use as many as possible.
+
+    Returns
+    -------
+        radiograph stacks, obs, dcs and omegas as numpy.ndarray
+
+    Notes
+    -----
+        There are two main signatures to load the data:
+        1. load_data(ct_files=ctfs, ob_files=obfs, dc_files=dcfs)
+        2. load_data(ct_dir=ctdir, ob_dir=obdir, dc_dir=dcdir)
+
+        The two signatures are mutually exclusive, and dc_files and dc_dir are optional
+        in both cases as some experiments do not have dark current measurements.
+
+        The fnmatch selectors are applicable in both signature, which help to down-select
+        files if needed. Default is set to "*", which selects everything.
+        Also, if ob_fnmatch and dc_fnmatch are set to "None" in the second signature call, the
```

Comment:
You discuss what happens if I do 
```python
load_data(ct_dir=ctdir, ob_dir=obdir, dc_dir=dcdir ob_fnmatch=None, dc_fnmatch=None)
``` 
but what happens if I do?
```python
load_data(ct_files=ctfs, ob_files=obfs, dc_files=dcfs, ob_fnmatch=None, dc_fnmatch=None)
```
Also, your discussion is a bit unclear for me. You say:

> if ob_fnmatch and dc_fnmatch are set to "None" in the second signature call, the data loader will attempt to read the metadata embedded in the ct file

but the second signature call receives `ct_dir` so what is "the ct file"? Do you mean each ct file in `ct_dir` that matches `ct_fnmatch`?

Reply:
if someone set it to none, it should be ignored as the internal will reset it to use all

Reply:
the default is already "*", and we cannot disallow None as it is used to trigger auto matching when using with x_dir.